include/libhal/usb.hpp

```diff
@@ -13,5 +13,588 @@
+   * function but with its size equal to the number of bytes read from the OUT
+   * endpoint. The size will be 0 if no more data was present in the endpoint.
+   */
+  [[nodiscard]] usize read(scatter_span<byte> p_buffer)
+  {
+    return driver_read(p_buffer);
+  }
+
+private:
+  virtual void driver_on_receive(
+    callback<void(on_receive_tag)> const& p_callback) = 0;
+  virtual usize driver_read(scatter_span<byte> p_buffer) = 0;
+};
+
+/**
+ * @brief USB Interrupt IN Endpoint Interface
+ *
+ * This class represents an interrupt IN endpoint of a USB device. Interrupt
+ * IN endpoints are used for small, time-sensitive data transfers from the
+ * device to the host.
+ *
+ * Use cases:
+ * - Sending periodic status updates
+ * - Transmitting small amounts of data with guaranteed latency
+ * - Ideal for devices like keyboards, mice, or game controllers
+ */
+struct interrupt_in_endpoint : public in_endpoint
+{};
+
+/**
+ * @brief USB Interrupt OUT Endpoint Interface
+ *
+ * This class represents an interrupt OUT endpoint of a USB device. Interrupt
+ * OUT endpoints are used for small, time-sensitive data transfers from the
+ * host to the device.
+ *
+ * Use cases:
+ * - Receiving periodic commands or settings
+ * - Handling small amounts of data with guaranteed latency
+ * - Ideal for devices that need quick responses to host commands
+ */
+struct interrupt_out_endpoint : public out_endpoint
+{};
+
+/**
+ * @brief USB Bulk IN Endpoint Interface
+ *
+ * This class represents a bulk IN endpoint of a USB device. Bulk IN endpoints
+ * are used for large, non-time-critical data transfers from the device to the
+ * host.
+ *
+ * Use cases:
+ * - Transferring large amounts of data
+ * - Sending data when timing is not critical
+ * - Ideal for devices like printers, scanners, or external storage
+ */
+struct bulk_in_endpoint : public in_endpoint
+{};
+
+/**
+ * @brief USB Bulk OUT Endpoint Interface
+ *
+ * This class represents a bulk OUT endpoint of a USB device. Bulk OUT
+ * endpoints are used for large, non-time-critical data transfers from the
+ * host to the device.
+ *
+ * Use cases:
+ * - Receiving large amounts of data
+ * - Handling data when timing is not critical
+ * - Ideal for devices like printers, scanners, or external storage
+ */
+struct bulk_out_endpoint : public out_endpoint
+{};
+
+template<class T>
+concept out_endpoint_type = std::is_base_of_v<out_endpoint, T> ||
+                            std::is_base_of_v<control_endpoint, T> ||
+                            std::is_base_of_v<bulk_out_endpoint, T> ||
+                            std::is_base_of_v<interrupt_out_endpoint, T>;
+
+template<class T>
+concept in_endpoint_type =
+  std::is_base_of_v<in_endpoint, T> || std::is_base_of_v<control_endpoint, T> ||
+  std::is_base_of_v<bulk_in_endpoint, T> ||
+  std::is_base_of_v<interrupt_in_endpoint, T>;
+
+/**
+ * @brief USB Setup Request packet definition
+ *
+ */
+struct setup
```

Comment:
Nit: Think it might be best to call this setup_packet just because this is a specific thing in the standard 
https://www.beyondlogic.org/usbnutshell/usb6.shtml#SetupPacket

Reply:
Sounds good!